src/common/types/int128_t.cpp

```diff
@@ -0,0 +1,650 @@
+    return result;
+}
+
+int128_t Int128_t::divModPositive(int128_t lhs, uint64_t rhs, uint64_t& remainder) {
+    int128_t result{};
+    result.low = 0;
+    result.high = 0;
+    remainder = 0;
+
+    for (uint8_t i = PositiveInt128BitsAmount(lhs); i > 0; i--) {
+        result = PositiveInt128LeftShift(result, 1);
+        remainder <<= 1;
+        if (PositiveInt128IsBitSet(lhs, i - 1)) {
+            remainder++;
+        }
+        if (remainder >= rhs) {
+            remainder -= rhs;
+            result.low++;
+            if (result.low == 0) {
+                result.high++;
+            }
+        }
+    }
+    return result;
+}
+
+std::string Int128_t::ToString(int128_t input) {
+    uint64_t remainder;
+    std::string result;
+    bool negative = input.high < 0;
+    if (negative) {
+        negateInPlace(input);
+    }
+    while (true) {
+        if (!input.low && !input.high) {
+            break;
+        }
+        input = divModPositive(input, 10, remainder);
+        result = std::string(1, '0' + remainder) + result;
+    }
+    if (result.empty()) {
+        result = "0";
+    }
+    return negative ? "-" + result : result;
+}
+
+//===============================================================================================
+// Arithmetic operators: ADD SUB
+//================================================================================================
+bool Int128_t::addInPlace(int128_t& lhs, int128_t rhs) {
+    int overflow = lhs.low + rhs.low < lhs.low;
+    if (rhs.high >= 0) {
+        if (lhs.high > INT64_MAX - rhs.high - overflow) {
+            return false;
+        }
+        lhs.high = lhs.high + rhs.high + overflow;
+    } else {
+        if (lhs.high < INT64_MIN - rhs.high - overflow) {
+            return false;
+        }
+        lhs.high = lhs.high + rhs.high + overflow;
+    }
+    lhs.low += rhs.low;
+    if (lhs.high == INT64_MIN && lhs.low == 0) {
+        return false;
+    }
+    return true;
+}
+
+bool Int128_t::subInPlace(int128_t& lhs, int128_t rhs) {
+    int underflow = lhs.low - rhs.low > lhs.low;
+    if (rhs.high >= 0) {
+        if (lhs.high < INT64_MIN + rhs.high + underflow) {
+            return false;
+        }
+        lhs.high = lhs.high - rhs.high - underflow;
+    } else {
+        if (lhs.high > INT64_MIN && lhs.high - 1 >= INT64_MAX + rhs.high + underflow) {
+            return false;
+        }
+        lhs.high = lhs.high - rhs.high - underflow;
+    }
+    lhs.low -= rhs.low;
+    if (lhs.high == INT64_MIN && lhs.low == 0) {
+        return false;
+    }
+    return true;
+}
+
+int128_t Int128_t::Add(int128_t lhs, const int128_t rhs) {
+    if (!addInPlace(lhs, rhs)) {
+        throw std::overflow_error("INT128 is out of range: cannot add.");
+    }
+    return lhs;
+}
+
+int128_t Int128_t::Sub(int128_t lhs, const int128_t rhs) {
+    if (!subInPlace(lhs, rhs)) {
+        throw std::overflow_error("INT128 is out of range: cannot subtract.");
```

Comment:
Instead of throwing c++ execptions, we should throw kuzu internal exception. Add a new exception type: overflowException